apollo-federation/src/sources/connect/validation/selection.rs

```diff
@@ -91,65 +93,105 @@ pub(super) fn validate_body_selection(
-        })?;
-
-    let selection = JSONSelection::parse(selection_str.as_str()).map_err(|err| Message {
-        code: Code::InvalidJsonSelection,
-        message: format!("{coordinate} is not a valid JSONSelection: {err}"),
-        locations: selection_node
-            .line_column_range(&schema.sources)
-            .into_iter()
-            .collect(),
-    })?;
-
+    // Ensure that the body selection is a valid JSON selection string
+    let selection_str = match GraphQLString::new(selection_node, &schema.sources) {
+        Ok(selection_str) => selection_str,
+        Err(_) => {
+            return vec![Message {
+                code: Code::GraphQLError,
+                message: format!("{coordinate} must be a string."),
+                locations: selection_node
+                    .line_column_range(&schema.sources)
+                    .into_iter()
+                    .collect(),
+            }]
+        }
+    };
+    let selection = match JSONSelection::parse(selection_str.as_str()) {
+        Ok(selection) => selection,
+        Err(err) => {
+            return vec![Message {
+                code: Code::InvalidJsonSelection,
+                message: format!("{coordinate} is not a valid JSONSelection: {err}"),
+                locations: selection_node
+                    .line_column_range(&schema.sources)
+                    .into_iter()
+                    .collect(),
+            }]
+        }
+    };
     if selection.is_empty() {
-        return Err(Message {
+        return vec![Message {
             code: Code::InvalidJsonSelection,
             message: format!("{coordinate} is empty"),
             locations: selection_node
                 .line_column_range(&schema.sources)
                 .into_iter()
                 .collect(),
-        });
+        }];
     }
-    let var_paths = selection.external_var_paths();
-    if var_paths.is_empty() {
-        return Err(Message {
+
+    // Validate the selection shape
+    let shape = selection.shape();
+    let expression_location = 0..selection_str.as_str().len();
+    let mut messages = Vec::new();
+    messages.extend(shape.errors().map(|err| {
+        Message {
             code: Code::InvalidJsonSelection,
-            message: format!("{coordinate} must contain at least one variable reference"),
-            locations: selection_node
-                .line_column_range(&schema.sources)
+            message: format!("In {coordinate}: {}", err.message),
+            locations: selection_str
+                .line_col_for_subslice(expression_location.narrow(err.range.as_ref()), schema)
                 .into_iter()
                 .collect(),
-        });
+        }
+    }));
+    match shape.case() {
+        ShapeCase::Name(name, _) if name == "$root" => messages.push(Message {
+            code: Code::InvalidJsonSelection,
+            message: format!("In {coordinate}: invalid body selection - body selections have no input to select from"),
```

Comment:
```suggestion
            message: format!("In {coordinate}: body selections cannot access the API response with `$`"),
```

or something. i think the word "input" here is potentially confusing

Reply:
Agreed - I'm not sure what the best way to convey this to customers is, but I'll think about it more.

Reply:
In URIs and headers, we just say that expressions must start with `$args`, `$this`, etc. (whatever is valid there). Technically a lie, because you can also use literals/methods, but if they're trying to use a bare path they _probably_ meant `$this`